### ISSUE TITLE
add log table for objects/records not currently auditable

### ISSUE THREAD

**Issue by ccaputo** [peeringdb/peeringdb#823]
For objects/records that don't have a history that can be used to track deletions or modifications, we should have an internal log table of some kind, which includes the user id of who performed an action.   @grizz, please flesh this out as appropriate.  @peeringdb/ac, support is welcome.

**Comment by arnoldnipper** [comment]
+1

**Comment by grizz** [comment]
+1 -- will discuss implementation ideas this week.

**Comment by peterhelmenstine** [comment]
+1 more data the better

**Comment by alexfguimaraes** [comment]
+1

**Comment by DarwinCosta** [comment]
+1.

Darwin-.

**Comment by mcmanuss8** [comment]
+1 
Would this cover users being added/removed from an organization? Just heard about a case where an organization was unsure how several users lost admin access and an audit log would have helped here.

**Comment by ccaputo** [comment]
> +1
> Would this cover users being added/removed from an organization? Just heard about a case where an organization was unsure how several users lost admin access and an audit log would have helped here.

If this data is not stored elsewhere, yes absolutely it should at least be in this new internal table.

**Comment by netravnen** [comment]
> If this data is not stored elsewhere, yes absolutely it should at least be in this new internal table.

I'm all in!

👍

**Comment by arnoldnipper** [comment]
> If this data is not stored elsewhere, yes absolutely it should at least be in this new internal table.

https://peeringdb.com/cp/peeringdb_server/user/nnnnn/history/ exists, but nothing is logged there

**Comment by ccaputo** [comment]
> > If this data is not stored elsewhere, yes absolutely it should at least be in this new internal table.
> 
> https://peeringdb.com/cp/peeringdb_server/user/nnnnn/history/ exists, but nothing is logged there. Keep in mind that due to GDPR we completely delete user-related data.

IIRC, we can maintain internally what we need to make sense of things - accountid, email addr, etc. - and disclose the practice in our Privacy Policy to comply with GDPR.  Thoughts?

**Comment by mcmanuss8** [comment]
That seems like a reasonable workaround, @ccaputo.

**Comment by arnoldnipper** [comment]
> IIRC, we can maintain internally what we need to make sense of things - accountid, email addr, etc. - and disclose the practice in our Privacy Policy to comply with GDPR. Thoughts?

Sure, we can log and maintain user-related data internally. However, when a user account is deleted these logs also have to  be deleted afaik.

**Comment by ccaputo** [comment]
> > IIRC, we can maintain internally what we need to make sense of things - accountid, email addr, etc. - and disclose the practice in our Privacy Policy to comply with GDPR. Thoughts?
> 
> Sure, we can log and maintain user-related data internally if we need them for operational purposes. However, when a user account is deleted these logs also have to be deleted afaik.

[Article 17 - Right to erasure (‘right to be forgotten’) - section 3(d)](https://gdpr-info.eu/art-17-gdpr/) makes the following exception:

---
- Paragraphs 1 and 2 shall not apply to the extent that processing is necessary:

  - for archiving purposes in the public interest, scientific or historical research purposes or statistical purposes in accordance with [Article 89(1)](https://gdpr-info.eu/art-89-gdpr/) in so far as the right referred to in paragraph 1 is likely to render impossible or seriously impair the achievement of the objectives of that processing; or
---

PeeringDB is organized and functions as a non-profit operating in the public interest.

That said, I'm all for retaining the absolute minimum needed to make sense of things, just short of shooting ourselves in the foot due to a lack of information. So, for example, maybe that means we maintain accountid and email-addr, but don't retain name/phone, etc.

Further, a normal account deletion is very different than a formal "Right to erasure" request.  Maybe when a "Right to erasure" request is received, we go one step further and nullify any email address in the audit log, while maintaining the accountid so there is some ability to audit the "Right to erasure" request years later.

**Comment by arnoldnipper** [comment]
I'm not sure I concur ... however IANAL ... as every one of us. Hence it might be worth to engage one.

**Comment by grizz** [comment]
Also not a lawyer, but I think we can reasonably interpret that, publicly disclose our policy, and do as @ccaputo said. Also agree that it might be worthwhile to run that by a real lawyer. :)

We can keep the log referencing UID who did the action, remove any other details after X days or a "Right to Erasure" request, and have a notes field for the audit log that explains the acting user's role.

**Comment by mcmanuss8** [comment]
Would we extend this to an external API and include all changes to all fields? I could see use cases for wanting a log of IP changes or members joining or leaving a given IX to trigger automation actions.

**Comment by ccaputo** [comment]
> Would we extend this to an external API and include all changes to all fields? I could see use cases for wanting a log of IP changes or members joining or leaving a given IX to trigger automation actions.

No.  I could see that as a different effort, but this is intended to assist us internally so that we are not blind.

**Comment by mcmanuss8** [comment]
> No. I could see that as a different effort, but this is intended to assist us internally so that we are not blind.

Fair - let's fork that idea to another issue... #915

**Comment by egfrank** [comment]
I think @ccaputo's [comment in issue 907](https://github.com/peeringdb/peeringdb/issues/907#issuecomment-756948373) contains good implementation details for this issue:

> Can you make it so when viewing the User, in addition to the "History" and "View on site" buttons, preceding "History" there will be an "Audit Log" button which then links to the audit log entries pertinent to the specific user?
> 
> For example: In addition to:
> 
> https://peeringdb.com/cp/peeringdb_server/user/[#]/change/
> there would be:
> 
> https://peeringdb.com/cp/peeringdb_server/user/[#]/log/
> which contains a log of any changes to the User object not listed on the history page.
> 
> Alternatively, or potentially more simply, can you just add non-Django Admin logging of User object changes to the existing User History?
> 
>

**Comment by leovegoda** [comment]
As this issue is currently in 3a Needs Implementation discussion, I'd like to understand if there's a link between it and #979

**Comment by arnoldnipper** [comment]
> As this issue is currently in 3a Needs Implementation discussion, I'd like to understand if there's a link between it and #979

No
